# Titanic Using Daimensions

This notebook uses data from the Titanic competition on Kaggle (https://www.kaggle.com/c/titanic/overview).

Kaggle's description of the competition:
"The sinking of the Titanic is one of the most infamous shipwrecks in history. On April 15, 1912, during her maiden voyage, the widely considered 'unsinkable' RMS Titanic sank after colliding with an iceberg. Unfortunately, there weren’t enough lifeboats for everyone onboard, resulting in the death of 1502 out of 2224 passengers and crew.
While there was some element of luck involved in surviving, it seems some groups of people were more likely to survive than others. In this challenge, we ask you to build a predictive model that answers the question: 'what sorts of people were more likely to survive?' using passenger data (ie name, age, gender, socio-economic class, etc)."

Goal: Make a predictor of survival from Titanic training data. We'll do this by using Daimensions to measure, build, and validate a predictor.

## 0. Getting Started

Because this is the very first tutorial, we'll go over how to install btc and get started. You can also see how to setup btc in the Daimensions Quickstart guide.

First, use the following link to download the installation script: https://download.brainome.net/btc-cli/btc-setup.sh. From the download directory, run the following bash command.

In [1]:
! sh btc-setup.sh

sh: btc-setup.sh: No such file or directory


The script will check that your operating system is supported, download the latest btc client to your machine and install it in /usr/local/bin. You will be prompted to enter the administrator password to install the software. 
*NOTE: After installation, make sure that “/usr/local/bin” is in your search path. *

Next, run the following command to wipe all cloud files. You will need your user credentials to login to DaimensionsTM. The first time you login, your license key will be downloaded automatically. Please use the default password that was provided to you. 

In [ ]:
! btc WIPE

To change your password, use the following bash command.

In [ ]:
! btc CHPASSWD

## 1. Get Measurements

Measuring our data before building a predictor is important in order to avoid mistakes and optimize our model. If we don't measure our data, we have no way of knowing whether the predictor we build will actually do what we want it to do when it sees new data that it wasn’t trained on. We'll probably build a model that is much larger than it needs to be, meaning our training and run times will probably be much longer than they need to be. We could end up in a situation where we just don’t know whether we have the right amount or right type of training data, even after extensive training and testing. Because of these reasons, it's best to measure our data beforehand. Not to mention, Daimensions will tell us about learnability, the generalization ratio, noise resilience, and all the standard accuracy and confusion figures. 
For more information, you can read the Daimensions How-to Guide and Glossary.

In [1]:
# Below is a clip of the training data:
! head data/titanic_train.csv
# For Windows command prompt:
# type titanic_train.csv | more

PassengerId,Cabin_Class,Name,Sex,Age,Sibling_Spouse,Parent_Children,Ticket_Number,Fare,Cabin_Number,Port_of_Embarkation,Survived
1,3,"Braund, Mr. Owen Harris",male,22,1,0,A/5 21171,7.25,,S,died
2,1,"Cumings, Mrs. John Bradley (Florence Briggs Thayer)",female,38,1,0,PC 17599,71.2833,C85,C,survived
3,3,"Heikkinen, Miss. Laina",female,26,0,0,STON/O2. 3101282,7.925,,S,survived
4,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35,1,0,113803,53.1,C123,S,survived
5,3,"Allen, Mr. William Henry",male,35,0,0,373450,8.05,,S,died
6,3,"Moran, Mr. James",male,,0,0,330877,8.4583,,Q,died
7,1,"McCarthy, Mr. Timothy J",male,54,0,0,17463,51.8625,E46,S,died
8,3,"Palsson, Master. Gosta Leonard",male,2,3,1,349909,21.075,,S,died
9,3,"Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)",female,27,0,2,347742,11.1333,,S,survived


As you can see from above, the target column (Survived) isn't the last column on the right. Because of this, we need to use '-target' so that Daimensions is looking at the correct target column for measuring and building a predictor.

In [2]:
# Measuring the training data:
! brainome -measureonly data/titanic_train.csv -target Survived 


Brainome Table Compiler v1.005-7-prod
Copyright (c) 2019-2021 Brainome, Inc. All Rights Reserved.
Licensed to:                 Alexander Makhratchev  (Evaluation)
Expiration Date:             2021-08-31   26 days left
Maximum File Size:           30 GB
Maximum Instances:           unlimited
Maximum Attributes:          unlimited
Maximum Classes:             unlimited
Connected to:                daimensions.brainome.ai  (local execution)

Command:
    btc -measureonly data/titanic_train.csv -target Survived

Start Time:                 08/05/2021, 17:11 PDT

Cleaning...done. 
Splitting into training and validation...done. 
Pre-training measurements...done. 


Pre-training Measurements
Data:
    Input:                      data/titanic_train.csv
    Target Column:              Survived
    Number of instances:        800
    Number of attributes:        11 out of 11
    Number of classes:            2

Class Balance:                
                            died: 61.50%
            

## 2. Build the Predictor

Because the learnability of the data (based on capacity progression and risk) is yellow, the how-to guide recommends to choose predictor with higher generalization and increase effort for best results. This means using a neural network with effort should work best. Here, I'm using '-f NN' to make the predictor a neural network. I'm also using '-o predict.py' to output the predictor as a python file. To increase the effort, I'm using '-e 10' for 10 times the effort. Again, we have to use '-target Survived' because the target column isn't the last one.

In [8]:
# Building the predictor and outputting it to 'titanic_predict.py':
! brainome -v -v -f NN data/titanic_train.csv -o titanic_predict.py -target Survived --yes


Brainome Table Compiler v1.005-7-prod
Copyright (c) 2019-2021 Brainome, Inc. All Rights Reserved.
Licensed to:                 Alexander Makhratchev  (Evaluation)
Expiration Date:             2021-08-31   26 days left
Maximum File Size:           30 GB
Maximum Instances:           unlimited
Maximum Attributes:          unlimited
Maximum Classes:             unlimited
Connected to:                daimensions.brainome.ai  (local execution)

Command:
    btc -v -v -f NN data/titanic_train.csv -o titanic_predict.py -target Survived --yes

Start Time:                 08/05/2021, 17:54 PDT

Cleaning...done. < 1s
Splitting into training and validation...done. 1s
Pre-training measurements...done. 2s


Pre-training Measurements
Data:
    Input:                      data/titanic_train.csv
    Target Column:              Survived
    Number of instances:        800
    Number of attributes:        11 out of 11
    Number of classes:            2

Class Balance:                
                  

## 3. Validate and Make Predictions

We've built our first predictor! Now it's time to put it to use. In the case of Titanic, we are given test data from Kaggle, where it's different from the training data and doesn't include 'Survival'. We can use the model we built to make predictions for the test data and submit it to Kaggle for its competition. In the following code, I'll save the model's prediction in 'titanic_prediction.csv'. You will see that the predictor appended the model's prediction of survival as the last column. 

In [9]:
# Using predictor on test data and saving it to 'titanic_prediction.csv':
! python3 titanic_predict.py data/titanic_test.csv > titanic_prediction.csv
! head titanic_prediction.csv

PassengerId,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked,Prediction
892,3,"Kelly, Mr. James",male,34.5,0,0,330911,7.8292,,Q,died
893,3,"Wilkes, Mrs. James (Ellen Needs)",female,47,1,0,363272,7,,S,died
894,2,"Myles, Mr. Thomas Francis",male,62,0,0,240276,9.6875,,Q,died
895,3,"Wirz, Mr. Albert",male,27,0,0,315154,8.6625,,S,died
896,3,"Hirvonen, Mrs. Alexander (Helga E Lindqvist)",female,22,1,1,3101298,12.2875,,S,died
897,3,"Svensson, Mr. Johan Cervin",male,14,0,0,7538,9.225,,S,died
898,3,"Connolly, Miss. Kate",female,30,0,0,330972,7.6292,,Q,died
899,2,"Caldwell, Mr. Albert Francis",male,26,1,1,248738,29,,S,died
900,3,"Abrahim, Mrs. Joseph (Sophie Halaut Easu)",female,18,0,0,2657,7.2292,,C,died


If you have validation data, or data that has the target column but wasn't used for training, you can use it to validate the accuracy of your predictor, as we will do. For this particular instance, I found an annotated version of the Titanic test data, 'titanic_validation.csv', and used it to validate our model.

In [7]:
# To validate:
! python3 titanic_predict.py -validate data/titanic_validate.csv

Classifier Type:                    Neural Network
System Type:                        2-way classifier

Accuracy:
    Best-guess accuracy:            61.25%
    Model accuracy:                 61.25% (49/80 correct)
    Improvement over best guess:    0.00% (of possible 38.75%)

Model capacity (MEC):               27 bits
Generalization ratio:               1.74 bits/bit

Confusion Matrix:

      Actual |   Predicted    
    ----------------------------
        died |      49        0
    survived |      31        0

Accuracy by Class:

      target | TP FP TN FN     TPR     TNR     PPV     NPV      F1      TS
    -------- | -- -- -- -- ------- ------- ------- ------- ------- -------
        died | 49 31  0  0 100.00%   0.00%  61.25%   0.00%  75.97%  61.25%
    survived |  0  0 49 31   0.00% 100.00%   0.00%  61.25%   0.00%   0.00%


From validating the predictor, we can see that it has 74.64% accuracy, 12.44% better than best-guess accuracy (which classifies all data points as the majority class). 

## 4. Improving Our Model

Our model did pretty well, but let's see if we can improve it. A column that contains a unique value in each row (for example a database key) will never contribute to generalization, so we shouldn't include database keys or other unique ID columns. We can remove these columns by using '-ignorecolumns'. We'll try ignoring columns: PassengerId, Name, Ticket, Cabin, Embarked, because they're all unique ID columns. We could also use '-rank' to rank columns by significance and only process contributing attributes.

### Ignorecolumns vs Rank:
There may be situations where domain knowledge suggests a better choice of features than -rank. If we know the data generative process, we can do better with -ignorecolumns than with -rank. Rank is also optimizing for quick clustering/decision tree. For neural networks, we may still wish to reduce input features, which can be done with pca, but at the cost of interpretability. Some applications may require the original features are used in which case pca isn't viable. Ignorecolumns can reduce features while maintaining interpretability and work better for neural networks than -rank may, but the burden of choosing the right columns to keep is now on us.

### Using -ignorecolumns:

In [9]:
# Using -ignorecolumns to make a better predictor:
! brainome -v -v -f NN data/titanic_train.csv -o titanic_predict_igcol.py -target Survived -ignorecolumns PassengerId,Name,Ticket_Number,Cabin_Number,Port_of_Embarkation -e 10 --yes


Brainome Table Compiler v1.005-7-prod
Copyright (c) 2019-2021 Brainome, Inc. All Rights Reserved.
Licensed to:                 Alexander Makhratchev  (Evaluation)
Expiration Date:             2021-08-31   22 days left
Maximum File Size:           30 GB
Maximum Instances:           unlimited
Maximum Attributes:          unlimited
Maximum Classes:             unlimited
Connected to:                daimensions.brainome.ai  (local execution)

Command:
    btc -v -v -f NN data/titanic_train.csv -o titanic_predict_igcol.py -target Survived -ignorecolumns PassengerId,Name,Ticket_Number,Cabin_Number,Port_of_Embarkation -e 10 --yes

Start Time:                 08/09/2021, 11:10 PDT

Cleaning...done. < 1s
Splitting into training and validation...done. 1s
Pre-training measurements...done. 11s


Pre-training Measurements
Data:
    Input:                      data/titanic_train.csv
    Target Column:              Survived
    Number of instances:        800
    Number of attributes:         6 out 

In [10]:
# Using the ignorecolumns predictor on test data and saving it to 'titanic_prediction_igcol.csv':
! python3 titanic_predict_igcol.py data/titanic_test.csv > titanic_prediction_igcol.csv
! head titanic_prediction_igcol.csv

PassengerId,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked,Prediction
892,3,"Kelly, Mr. James",male,34.5,0,0,330911,7.8292,,Q,died
893,3,"Wilkes, Mrs. James (Ellen Needs)",female,47,1,0,363272,7,,S,died
894,2,"Myles, Mr. Thomas Francis",male,62,0,0,240276,9.6875,,Q,died
895,3,"Wirz, Mr. Albert",male,27,0,0,315154,8.6625,,S,died
896,3,"Hirvonen, Mrs. Alexander (Helga E Lindqvist)",female,22,1,1,3101298,12.2875,,S,died
897,3,"Svensson, Mr. Johan Cervin",male,14,0,0,7538,9.225,,S,died
898,3,"Connolly, Miss. Kate",female,30,0,0,330972,7.6292,,Q,survived
899,2,"Caldwell, Mr. Albert Francis",male,26,1,1,248738,29,,S,died
900,3,"Abrahim, Mrs. Joseph (Sophie Halaut Easu)",female,18,0,0,2657,7.2292,,C,survived


As we wanted, -ignorecolumns removed the PassengerId, Name, Ticket, Cabin, and Embarked attributes. Next, we can use -validate to check the accuracy of our new predictor.

In [11]:
# Validating the -ignorecolumns predictor
! python3 titanic_predict_igcol.py -validate data/titanic_validate.csv

Classifier Type:                    Neural Network
System Type:                        2-way classifier

Accuracy:
    Best-guess accuracy:            61.25%
    Model accuracy:                 80.00% (64/80 correct)
    Improvement over best guess:    18.75% (of possible 38.75%)

Model capacity (MEC):               17 bits
Generalization ratio:               3.62 bits/bit

Confusion Matrix:

      Actual |   Predicted    
    ----------------------------
        died |      43        6
    survived |      10       21

Accuracy by Class:

      target | TP FP TN FN     TPR     TNR     PPV     NPV      F1      TS
    -------- | -- -- -- -- ------- ------- ------- ------- ------- -------
        died | 43 10 21  6  87.76%  67.74%  81.13%  77.78%  84.31%  72.88%
    survived | 21  6 43 10  67.74%  87.76%  77.78%  81.13%  72.41%  56.76%


Using -ignorecolumns has improved our accuracy to 80.00% from 74.64% originally.

### Using -rank:

In [21]:
# Using -rank to make a better predictor:
! brainome -v -v -f NN data/titanic_train.csv -o titanic_predict_rank.py -target Survived -rank --yes -e 10


Brainome Table Compiler v1.005-7-prod
Copyright (c) 2019-2021 Brainome, Inc. All Rights Reserved.
Licensed to:                 Alexander Makhratchev  (Evaluation)
Expiration Date:             2021-08-31   26 days left
Maximum File Size:           30 GB
Maximum Instances:           unlimited
Maximum Attributes:          unlimited
Maximum Classes:             unlimited
Connected to:                daimensions.brainome.ai  (local execution)

Command:
    btc -v -v -f NN data/titanic_train.csv -o titanic_predict_rank.py -target Survived -rank --yes -e 10

Start Time:                 08/05/2021, 20:02 PDT

Cleaning...done. < 1s
Ranking attributes...done. < 1s

Attribute Ranking:
    Columns selected:           Sex, Sibling_Spouse, Parent_Children, Cabin_Class
    Risk of coincidental column correlation:    0.0%
    Ignoring columns:           PassengerId, Name, Age, Ticket_Number, Fare, Cabin_Number, Port_of_Embarkation
    Test Accuracy Progression:
                                       

In [22]:
# Using the rank predictor on test data and saving it to 'titanic_prediction_rank.csv':
! python3 titanic_predict_rank.py data/titanic_test.csv > titanic_prediction_rank.csv
! head titanic_prediction_rank.csv

PassengerId,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked,Prediction
892,3,"Kelly, Mr. James",male,34.5,0,0,330911,7.8292,,Q,died
893,3,"Wilkes, Mrs. James (Ellen Needs)",female,47,1,0,363272,7,,S,died
894,2,"Myles, Mr. Thomas Francis",male,62,0,0,240276,9.6875,,Q,died
895,3,"Wirz, Mr. Albert",male,27,0,0,315154,8.6625,,S,died
896,3,"Hirvonen, Mrs. Alexander (Helga E Lindqvist)",female,22,1,1,3101298,12.2875,,S,died
897,3,"Svensson, Mr. Johan Cervin",male,14,0,0,7538,9.225,,S,died
898,3,"Connolly, Miss. Kate",female,30,0,0,330972,7.6292,,Q,survived
899,2,"Caldwell, Mr. Albert Francis",male,26,1,1,248738,29,,S,died
900,3,"Abrahim, Mrs. Joseph (Sophie Halaut Easu)",female,18,0,0,2657,7.2292,,C,survived


You can see that -rank decided to only look at the columns 'Sex','Parch' (Parent/child), and 'Fare'. This makes a lot of sense that the determining factors for survival on the Titanic were sex, how many parents or children they had on board, and how much their fare was. Seeing what attributes -rank chooses gives us powerful insight into understanding our data and its correlations.

In [6]:
# Validating the -rank predictor
! python3 titanic_predict_rank.py -validate data/titanic_validate.csv

Classifier Type:                    Neural Network
System Type:                        2-way classifier

Accuracy:
    Best-guess accuracy:            61.25%
    Model accuracy:                 80.00% (64/80 correct)
    Improvement over best guess:    18.75% (of possible 38.75%)

Model capacity (MEC):               25 bits
Generalization ratio:               2.47 bits/bit

Confusion Matrix:

      Actual |   Predicted    
    ----------------------------
        died |      46        3
    survived |      13       18

Accuracy by Class:

      target | TP FP TN FN     TPR     TNR     PPV     NPV      F1      TS
    -------- | -- -- -- -- ------- ------- ------- ------- ------- -------
        died | 46 13 18  3  93.88%  58.06%  77.97%  85.71%  85.19%  74.19%
    survived | 18  3 46 13  58.06%  93.88%  85.71%  77.97%  69.23%  52.94%


With -rank, our accuracy is 80.00%, again, an improvement over our original 74.64%.

## 5 Using a Decision Tree

Now we will be using a different model that the Brainome Table Compiler has to offer: a decision tree. We will also be using the -rank command in order to select the most correlated columns with the survival column. This will prevent us from overfitting our model with the training data.

In [29]:
! brainome data/titanic_train.csv -rank -f DT -o titanic_predict_DT.py 


Brainome Table Compiler v1.005-7-prod
Copyright (c) 2019-2021 Brainome, Inc. All Rights Reserved.
Licensed to:                 Alexander Makhratchev  (Evaluation)
Expiration Date:             2021-08-31   25 days left
Maximum File Size:           30 GB
Maximum Instances:           unlimited
Maximum Attributes:          unlimited
Maximum Classes:             unlimited
Connected to:                daimensions.brainome.ai  (local execution)

Command:
    btc data/titanic_train.csv -rank -f DT -o titanic_predict_DT.py

Start Time:                 08/06/2021, 19:07 PDT

Cleaning...done. 
Ranking attributes...done. 

Attribute Ranking:
    Columns selected:           Sex, Sibling_Spouse, Parent_Children, Cabin_Class
    Risk of coincidental column correlation:    0.0%
    
    Test Accuracy Progression:
                                          Sex :   78.75%
                               Sibling_Spouse :   79.62% change   +0.88%
                              Parent_Children :   80.25% cha

One think we notice is that the training and validation score are very similar to one another and this can be attributed to the fact that we did not overfit the training data.

In [30]:
# Using the Decision Tree predictor on test data and saving it to 'titanic_prediction_DT.csv':
! python3 titanic_predict_DT.py data/titanic_test.csv > titanic_prediction_DT.csv
! head titanic_prediction_DT.csv

PassengerId,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked,Prediction
892,3,"Kelly, Mr. James",male,34.5,0,0,330911,7.8292,,Q,died
893,3,"Wilkes, Mrs. James (Ellen Needs)",female,47,1,0,363272,7,,S,survived
894,2,"Myles, Mr. Thomas Francis",male,62,0,0,240276,9.6875,,Q,died
895,3,"Wirz, Mr. Albert",male,27,0,0,315154,8.6625,,S,died
896,3,"Hirvonen, Mrs. Alexander (Helga E Lindqvist)",female,22,1,1,3101298,12.2875,,S,survived
897,3,"Svensson, Mr. Johan Cervin",male,14,0,0,7538,9.225,,S,died
898,3,"Connolly, Miss. Kate",female,30,0,0,330972,7.6292,,Q,survived
899,2,"Caldwell, Mr. Albert Francis",male,26,1,1,248738,29,,S,died
900,3,"Abrahim, Mrs. Joseph (Sophie Halaut Easu)",female,18,0,0,2657,7.2292,,C,survived


In [13]:
# Validating the DT predictor
! python3 titanic_predict_DT.py -validate data/titanic_validate.csv

Classifier Type:                    Decision Tree
System Type:                        2-way classifier

Accuracy:
    Best-guess accuracy:            61.25%
    Model accuracy:                 81.25% (65/80 correct)
    Improvement over best guess:    20.00% (of possible 38.75%)

Model capacity (MEC):               1 bits
Generalization ratio:               62.61 bits/bit

Confusion Matrix:

      Actual |   Predicted    
    ----------------------------
        died |      45        4
    survived |      11       20

Accuracy by Class:

      target | TP FP TN FN     TPR     TNR     PPV     NPV      F1      TS
    -------- | -- -- -- -- ------- ------- ------- ------- ------- -------
        died | 45 11 20  4  91.84%  64.52%  80.36%  83.33%  85.71%  75.00%
    survived | 20  4 45 11  64.52%  91.84%  83.33%  80.36%  72.73%  57.14%


Our accuracy has no gone up to 81.25% on the validation data.

## 6. Random Forest Model

Now we are going to try the third and final model that the Brainome Table Compiler has to offer: random forest classifier. A majority of the time this model will have the best result on the validation set and is able to trained fairly quickly eventhough it is more intensive than a decision tree. 

In [32]:
! brainome data/titanic_train.csv -rank -f RF -o titanic_predict_RF.py 


Brainome Table Compiler v1.005-7-prod
Copyright (c) 2019-2021 Brainome, Inc. All Rights Reserved.
Licensed to:                 Alexander Makhratchev  (Evaluation)
Expiration Date:             2021-08-31   25 days left
Maximum File Size:           30 GB
Maximum Instances:           unlimited
Maximum Attributes:          unlimited
Maximum Classes:             unlimited
Connected to:                daimensions.brainome.ai  (local execution)

Command:
    btc data/titanic_train.csv -rank -f RF -o titanic_predict_RF.py

Start Time:                 08/06/2021, 19:21 PDT

Cleaning...done. 
Ranking attributes...done. 

Attribute Ranking:
    Columns selected:           Sex, Sibling_Spouse, Parent_Children, Cabin_Class
    Risk of coincidental column correlation:    0.0%
    
    Test Accuracy Progression:
                                          Sex :   78.75%
                               Sibling_Spouse :   79.62% change   +0.88%
                              Parent_Children :   80.25% cha

In [34]:
# Using the Random Forest predictor on test data and saving it to 'titanic_prediction_RF.csv':
! python3 titanic_predict_RF.py data/titanic_test.csv > titanic_prediction_RF.csv
! head titanic_prediction_RF.csv

PassengerId,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked,Prediction
892,3,"Kelly, Mr. James",male,34.5,0,0,330911,7.8292,,Q,died
893,3,"Wilkes, Mrs. James (Ellen Needs)",female,47,1,0,363272,7,,S,survived
894,2,"Myles, Mr. Thomas Francis",male,62,0,0,240276,9.6875,,Q,died
895,3,"Wirz, Mr. Albert",male,27,0,0,315154,8.6625,,S,died
896,3,"Hirvonen, Mrs. Alexander (Helga E Lindqvist)",female,22,1,1,3101298,12.2875,,S,survived
897,3,"Svensson, Mr. Johan Cervin",male,14,0,0,7538,9.225,,S,died
898,3,"Connolly, Miss. Kate",female,30,0,0,330972,7.6292,,Q,survived
899,2,"Caldwell, Mr. Albert Francis",male,26,1,1,248738,29,,S,died
900,3,"Abrahim, Mrs. Joseph (Sophie Halaut Easu)",female,18,0,0,2657,7.2292,,C,survived


In [14]:
# Validating the DT predictor
! python3 titanic_predict_RF.py -validate data/titanic_validate.csv

Classifier Type:                    Random Forest
System Type:                        2-way classifier

Accuracy:
    Best-guess accuracy:            61.25%
    Model accuracy:                 80.00% (64/80 correct)
    Improvement over best guess:    18.75% (of possible 38.75%)

Model capacity (MEC):               17 bits
Generalization ratio:               3.62 bits/bit

Confusion Matrix:

      Actual |   Predicted    
    ----------------------------
        died |      45        4
    survived |      12       19

Accuracy by Class:

      target | TP FP TN FN     TPR     TNR     PPV     NPV      F1      TS
    -------- | -- -- -- -- ------- ------- ------- ------- ------- -------
        died | 45 12 19  4  91.84%  61.29%  78.95%  82.61%  84.91%  73.77%
    survived | 19  4 45 12  61.29%  91.84%  82.61%  78.95%  70.37%  54.29%


Our accuracy is 80.00% on the validation set.

## 7. Next Steps

Success! We've built our first predictor and used it to make predictions on the Titanic test data. From here, we can use our model on any new Titanic data or use other control options to try to improve our results even more.
To check out some of the other control options, use '-h' to see the full list. You can also check out Brainome's How-to Guide and Glossary for more information.

In [11]:
! brainome -h

usage: brainome [-h] [-version] [-headerless] [-target TARGET]
                [-ignorecolumns IGNORECOLUMNS] [-rank [ATTRIBUTERANK]]
                [-measureonly] [-f FORCEMODEL] [-nosplit] [-split FORCESPLIT]
                [-nsamples NSAMPLES] [-ignoreclasses IGNORELABELS]
                [-usecolumns IMPORTANTCOLUMNS] [-o OUTPUT] [-v] [-q] [-y]
                [-e EFFORT] [-biasmeter] [-novalidation] [-balance]
                [-O OPTIMIZE] [-nofun] [-modelonly]
                input [input ...]

Brainome Table Compiler (tm)  v1.005-7-prod

Required arguments:
  input                 Table as CSV files and/or URLs or Command above

Optional arguments:
  -h                    show this help message and exit
  -version              show program's version number and exit

Basic options:
  -headerless           Headerless CSV input file.
  -target TARGET        Specify target column by name or number. Default: last column of table.
  -ignorecolumns IGNORECOLUMNS
                     